File: src/typing/exists_marker.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Flow_polymorphic_ast_mapper
module Ast = Flow_ast

class marker cx =
  object (this)
    inherit [ALoc.t, ALoc.t * Type.t, ALoc.t, ALoc.t * Type.t] mapper as super

    val mutable cond = false

    val mutable predicate = false

    method on_type_annot x = x

    method on_loc_annot x = x

    method push : 'a 'r. set:('a -> unit) -> get:(unit -> 'a) -> 'a -> 'r Lazy.t -> 'r =
      fun ~set ~get v r ->
        let init = get () in
        set v;
        let res = Lazy.force r in
        set init;
        res

    method push_pred = this#push ~set:(fun x -> predicate <- x) ~get:(fun () -> predicate)

    method! expression expr = this#base_expression false expr

    method! predicate_expression expr = this#base_expression true expr

    method base_expression c expr =
      let open Ast.Expression in
      cond <- c;
      begin
        if c then
          match expr with
          | ( (loc, ty),
              OptionalMember
                { OptionalMember.member = { Member.property = Member.PropertyExpression _; _ }; _ }
            )
          | (_, OptionalMember { OptionalMember.filtered_out = (loc, ty); _ })
          | ((loc, ty), Member _)
          | ((_, ty), Assignment { Assignment.left = ((loc, _), Ast.Pattern.Identifier _); _ }) ->
            Context.add_exists_check cx loc ty
          | ((loc, ty), _) when Base.Option.is_some (Refinement.Keys.key ~allow_optional:false expr)
            ->
            Context.add_exists_check cx loc ty
          | _ -> ()
      end;
      super#expression expr

    method! type_predicate (pred : (_, _) Ast.Type.Predicate.t) =
      let open Ast.Type.Predicate in
      let (annot, { kind; comments = _ }) = pred in
      let _annot' = this#on_loc_annot annot in
      let _kind' =
        match kind with
        | Inferred -> kind
        | Declared expr -> Declared (this#base_expression cond expr)
      in
      pred

    method! return (stmt : (_, _) Ast.Statement.Return.t) =
      let open Ast.Statement.Return in
      let { argument; comments = _; return_out = _ } = stmt in
      let _argument' = Base.Option.map ~f:(this#base_expression predicate) argument in
      stmt

    method! function_ ({ Ast.Function.predicate; _ } as fn) =
      this#push_pred (Base.Option.is_some predicate) (lazy (super#function_ fn))

    method! call annot (expr : (_, _) Ast.Expression.Call.t) =
      let open Ast.Expression.Call in
      let { callee; targs; arguments; comments = _ } = expr in
      match (callee, targs, arguments) with
      | ( (_, Flow_ast.Expression.Identifier (_, { Flow_ast.Identifier.name = "invariant"; _ })),
          None,
          ( _,
            {
              Ast.Expression.ArgList.arguments = Ast.Expression.Expression conditional :: arguments;
              comments = _;
            }
          )
        ) ->
        let _ = this#expression callee in
        let _ = this#base_expression true conditional in
        let _ = Base.List.map ~f:this#expression_or_spread arguments in
        expr
      | _ -> super#call annot expr

    method! logical (expr : (_, _) Ast.Expression.Logical.t) =
      let open Ast.Expression.Logical in
      let { operator; left; right; comments = _ } = expr in
      let c =
        match operator with
        | Ast.Expression.Logical.(And | Or) -> true
        | Ast.Expression.Logical.NullishCoalesce -> false
      in
      let cur_c = cond in
      let _left' = this#base_expression c left in
      let _right' = this#base_expression cur_c right in
      expr

    method! assignment (expr : (_, _) Ast.Expression.Assignment.t) =
      let open Ast.Expression.Assignment in
      let { operator; left; right; comments = _ } = expr in
      let left_expr =
        match left with
        | (lhs_loc, Ast.Pattern.Identifier { Ast.Pattern.Identifier.name; _ }) ->
          Some (lhs_loc, Ast.Expression.Identifier name)
        | (lhs_loc, Ast.Pattern.Expression (_, Ast.Expression.Member mem)) ->
          Some (lhs_loc, Ast.Expression.Member mem)
        | _ -> None
      in
      match (operator, left_expr) with
      | (Some Ast.Expression.Assignment.(AndAssign | OrAssign), Some left_expr) ->
        let _ = this#base_expression true left_expr in
        let _ = this#expression right in
        expr
      | _ -> super#assignment expr

    method! unary_expression (expr : (_, _) Ast.Expression.Unary.t) =
      let open Ast.Expression.Unary in
      let { argument; operator; comments = _ } = expr in
      match operator with
      | Not ->
        let _ = this#base_expression cond argument in
        expr
      | _ -> super#unary_expression expr
  end

let mark cx ast =
  let marker = new marker cx in
  let _ = marker#program ast in
  ()

File: rust_port/crates/flow_typing_statement/src/exists_marker.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::ops::Deref;

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_parser::ast;
use flow_parser::ast::expression;
use flow_parser::ast::expression::ExpressionInner;
use flow_parser::ast_utils;
use flow_parser::ast_visitor;
use flow_parser::ast_visitor::AstVisitor;
use flow_typing_context::Context;
use flow_typing_type::type_::Type;

use crate::refinement;

struct Marker<'cx> {
    cond: bool,
    predicate: bool,
    cx: &'cx Context,
}

impl<'cx> Marker<'cx> {
    fn new(cx: &'cx Context) -> Self {
        Marker {
            cond: false,
            predicate: false,
            cx,
        }
    }

    fn push_pred<R>(&mut self, v: bool, f: impl FnOnce(&mut Self) -> R) -> R {
        let init = self.predicate;
        self.predicate = v;
        let res = f(self);
        self.predicate = init;
        res
    }

    //     method base_expression c expr =
    fn base_expression<'ast>(
        &mut self,
        c: bool,
        expr: &'ast expression::Expression<ALoc, (ALoc, Type)>,
    ) where
        'cx: 'ast,
    {
        self.cond = c;
        if c {
            match expr.deref() {
                ExpressionInner::OptionalMember { loc, inner }
                    if matches!(
                        inner.member.property,
                        expression::member::Property::PropertyExpression(_)
                    ) =>
                {
                    let (loc, ty) = loc;
                    self.cx.add_exists_check(loc.dupe(), ty.dupe());
                }
                ExpressionInner::OptionalMember { inner, .. } => {
                    let (loc, ty) = &inner.filtered_out;
                    self.cx.add_exists_check(loc.dupe(), ty.dupe());
                }
                ExpressionInner::Member { loc, .. } => {
                    let (loc, ty) = loc;
                    self.cx.add_exists_check(loc.dupe(), ty.dupe());
                }
                ExpressionInner::Assignment { loc, inner } => {
                    let (_, ty) = loc;
                    let (unwrapped_left, _) = ast_utils::unwrap_nonnull_lhs(&inner.left);
                    match &*unwrapped_left {
                        ast::pattern::Pattern::Identifier { loc: pat_loc, .. } => {
                            let (pat_aloc, _) = pat_loc;
                            self.cx.add_exists_check(pat_aloc.dupe(), ty.dupe());
                        }
                        _ => {}
                    }
                }
                _ => {
                    if refinement::keys::key(false, expr).is_some() {
                        let (loc, ty) = expr.loc();
                        self.cx.add_exists_check(loc.dupe(), ty.dupe());
                    }
                }
            }
        }
        let Ok(()) = ast_visitor::expression_default(self, expr);
    }
}

impl<'ast, 'cx: 'ast> AstVisitor<'ast, ALoc, (ALoc, Type), &'ast ALoc, !> for Marker<'cx> {
    fn normalize_loc(loc: &'ast ALoc) -> &'ast ALoc {
        loc
    }

    fn normalize_type(_type: &'ast (ALoc, Type)) -> &'ast ALoc {
        &_type.0
    }

    fn expression(
        &mut self,
        expr: &'ast expression::Expression<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        self.base_expression(false, expr);
        Ok(())
    }

    fn predicate_expression(
        &mut self,
        expr: &'ast expression::Expression<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        self.base_expression(true, expr);
        Ok(())
    }

    fn predicate(
        &mut self,
        pred: &'ast ast::types::Predicate<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let ast::types::Predicate {
            loc: _,
            kind,
            comments: _,
        } = pred;
        match kind {
            ast::types::PredicateKind::Inferred => {}
            ast::types::PredicateKind::Declared(expr) => {
                self.base_expression(self.cond, expr);
            }
        }
        Ok(())
    }

    fn return_(
        &mut self,
        _loc: &'ast ALoc,
        stmt: &'ast ast::statement::Return<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let ast::statement::Return {
            argument,
            comments: _,
            return_out: _,
        } = stmt;
        if let Some(argument) = argument {
            self.base_expression(self.predicate, argument);
        }
        Ok(())
    }

    fn function_(
        &mut self,
        loc: &'ast ALoc,
        fn_: &'ast ast::function::Function<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let has_predicate = fn_.predicate.is_some();
        self.push_pred(has_predicate, |this| {
            let Ok(()) = ast_visitor::function_default(this, loc, fn_);
        });
        Ok(())
    }

    fn call(
        &mut self,
        loc: &'ast (ALoc, Type),
        expr: &'ast expression::Call<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let expression::Call {
            callee,
            targs,
            arguments,
            comments: _,
        } = expr;
        let is_invariant = matches!(
            callee.deref(),
            ExpressionInner::Identifier { inner, .. }
            if inner.name.as_str() == "invariant"
        );
        if is_invariant && targs.is_none() {
            if let Some((expression::ExpressionOrSpread::Expression(conditional), rest_args)) =
                arguments.arguments.split_first()
            {
                self.expression(callee)?;
                self.base_expression(true, conditional);
                for arg in rest_args {
                    self.expression_or_spread(arg)?;
                }
                return Ok(());
            }
        }
        let Ok(()) = ast_visitor::call_default(self, loc, expr);
        Ok(())
    }

    fn logical(
        &mut self,
        _loc: &'ast (ALoc, Type),
        expr: &'ast expression::Logical<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let expression::Logical {
            operator,
            left,
            right,
            comments: _,
        } = expr;
        let c = match operator {
            expression::LogicalOperator::And | expression::LogicalOperator::Or => true,
            expression::LogicalOperator::NullishCoalesce => false,
        };
        let cur_c = self.cond;
        self.base_expression(c, left);
        self.base_expression(cur_c, right);
        Ok(())
    }

    fn assignment(
        &mut self,
        _loc: &'ast (ALoc, Type),
        expr: &'ast expression::Assignment<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let expression::Assignment {
            operator,
            left,
            right,
            comments: _,
        } = expr;
        let (unwrapped_left, _) = ast_utils::unwrap_nonnull_lhs(left);
        let left_expr: Option<expression::Expression<ALoc, (ALoc, Type)>> = match &*unwrapped_left {
            ast::pattern::Pattern::Identifier {
                loc: lhs_loc,
                inner,
            } => Some(expression::Expression::new(ExpressionInner::Identifier {
                loc: lhs_loc.dupe(),
                inner: inner.name.dupe(),
            })),
            ast::pattern::Pattern::Expression {
                loc: lhs_loc,
                inner,
            } => match &***inner {
                ExpressionInner::Member { inner: mem, .. } => {
                    Some(expression::Expression::new(ExpressionInner::Member {
                        loc: lhs_loc.dupe(),
                        inner: mem.dupe(),
                    }))
                }
                _ => None,
            },
            _ => None,
        };
        match (operator, left_expr) {
            (
                Some(
                    expression::AssignmentOperator::AndAssign
                    | expression::AssignmentOperator::OrAssign,
                ),
                Some(left_expr),
            ) => {
                self.base_expression(true, &left_expr);
                self.expression(right)?;
            }
            _ => {
                let Ok(()) = ast_visitor::assignment_default(self, _loc, expr);
            }
        }
        Ok(())
    }

    fn unary_expression(
        &mut self,
        _loc: &'ast (ALoc, Type),
        expr: &'ast expression::Unary<ALoc, (ALoc, Type)>,
    ) -> Result<(), !> {
        let expression::Unary {
            argument,
            operator,
            comments: _,
        } = expr;
        match operator {
            expression::UnaryOperator::Not => {
                self.base_expression(self.cond, argument);
            }
            _ => {
                let Ok(()) = ast_visitor::unary_expression_default(self, _loc, expr);
            }
        }
        Ok(())
    }
}

pub fn mark(cx: &Context, ast: &ast::Program<ALoc, (ALoc, Type)>) {
    let mut marker = Marker::new(cx);
    let Ok(()) = marker.program(ast);
}
